torch/lib/c10d/test/ProcessGroupGlooTest.cpp

```diff
@@ -268,87 +269,67 @@ void testSend(const std::string& path) {
   auto tests = CollectiveTest::initialize(path, size);
 
   constexpr uint64_t tag = 0x1337;
-
-  // Create a sender thread and ensure that sendWork can be aborted.
-  std::thread senderThread([&]() {
-    auto selfRank = 0;
-    auto destRank = 1;
-
-    auto& pg = tests[selfRank].getProcessGroup();
-
-    std::vector<at::Tensor> tensors = {
-        at::ones({16, 16}),
-    };
-    // Ensure that we can abort SendWork
-    auto work = pg.send(tensors, destRank, tag);
-    bool sendCompleted;
-    std::thread waitSendWorkThread([&]() { sendCompleted = work->wait(); });
-    work->abort();
-    waitSendWorkThread.join();
-    TORCH_CHECK(!sendCompleted);
-
-    // Ensure that future waits can complete successfully.
-    work = pg.send(tensors, destRank, tag);
-    sendCompleted = work->wait();
-    TORCH_CHECK(sendCompleted);
-  });
-
-  // Helper receiver thread to simulate a real send/recv pair.
-  std::thread receiverThread([&]() {
+  // test that waiting for work to be sent can be aborted successfully.
+  auto selfRank = 0;
+  auto dstRank = 1;
+  std::vector<at::Tensor> tensors = {
+      at::ones({16, 16}),
+  };
+  auto& pg = tests[selfRank].getProcessGroup();
+  auto sendWork = pg.send(tensors, dstRank, tag);
+  bool sendCompleted;
+  std::thread waitSendThreadAbort([&]() { sendCompleted = sendWork->wait(); });
+  sendWork->abort();
+  // Block until the sendWork gets successfully aborted
+  waitSendThreadAbort.join();
+  ASSERT_FALSE(sendCompleted);
+
+  // Now create a separate sender thread to ensure that future waitsends can
+  // complete successfully.
+
+  // Helper receiver to simulate a real recv/send pair
+  std::thread recvThread([&]() {
     auto selfRank = 1;
     auto srcRank = 0;
-
     auto& pg = tests[selfRank].getProcessGroup();
-
     std::vector<at::Tensor> tensors = {
         at::ones({16, 16}),
     };
 
-    // Receive for first send.
-    auto work = pg.recv(tensors, srcRank, tag);
-    work->wait();
-
-    // Receive for 2nd send.
-    work = pg.recv(tensors, srcRank, tag);
-    work->wait();
+    auto recvWork = pg.recv(tensors, srcRank, tag);
+    recvWork->wait();
   });
 
-  senderThread.join();
-  receiverThread.join();
+  // Sender thread
+  std::thread sendThread([&]() { sendCompleted = sendWork->wait(); });
```

Comment:
How does this work? Didn't we abort `sendWork` on line 282? Don't we need to create a new `sendWork` object here?

Reply:
@pritamdamania87 If you look at the implementation of `ProcessGroup.cpp`, it is actually the underlying gloo `buffer_` that can be aborted (see the tests in https://github.com/facebookincubator/gloo/pull/232/files which this is designed to mirror). If we create a new `sendWork` here via `pg.send()`, it will create a new buffer internally in `ProcessGroup`, but if we have a successful receive, it will be processed by the previous `sendWork` object (i.e. it's underlying `buffer_`). So if we created a new `sendWork` here the old one will get a completion and this one won't, so the test will fail. If we want to change this behavior so that we do need to create a new `sendWork` object, then that'll require changes to `ProcessGroup` and/or gloo's `unbound_buffer`. In terms of implementation, I'm not sure which is more correct

Reply:
This behavior does look a little weird. It means that even if the sender has aborted the send, and then come back to check the result of `wait()`, it could still return `True`? This is `ProcessGroupGloo::SendWork`, right? Is it possible to destroy the `buffer_` when abort is called? Will that prevent this from happening?

Reply:
> This behavior does look a little weird. It means that even if the sender has aborted the send, and then come back to check the result of `wait()`, it could still return `True`? 

Right, this is possible. It's not really "aborting the send" though, it's more like we aborted waiting for the send (write) to be completed. This doesn't entail that we're cancelling the send, we're just no longer blocking waiting for it. We've already scheduled the write (asynchronously or not). It could still be completed.

Likewise for the `recv` side, we've only aborted blocking on waiting for the receive to be completed, we can't really (it seems to me) cancel receiving the data itself (since we've already called `buf->recv()`. 

> This is `ProcessGroupGloo::SendWork`, right? Is it possible to destroy the `buffer_` when abort is called? Will that prevent this from happening?

I can look into this.

Reply:
Thanks. This shouldn't block us from landing this PR. This fix is great, and `abort` is not used by any applications other than us I think, because below is how we expose abort currently (will remove it):

```c++
          .def(
              "abort",
              &::c10d::ProcessGroup::barrier,
              py::arg("opts") = ::c10d::BarrierOptions(),
              py::call_guard<py::gil_scoped_release>())
```

For now, maybe we can add some comment to the `abort()` API describing the current behavior. The fix for destroying buffer can come in in followup PRs.

Reply:
@mrshenli I'm pretty sure that this is a different abort function, the one that this tests is the `ProcessGroup::Work::abort()` function (https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroup.cpp#L47 , implemented by sublcasses `SendWork` and `RecvWork`) (which is also only used by us in ProcessGroupAgent), the one that you referred to seems to be calling `Processgroup::barrier()`.